wp-includes/class-wp-service-workers.php

```diff
@@ -522,30 +480,293 @@ public function register_precached_routes( $routes ) {
+			$url = apply_filters( 'script_loader_src', $url, $handle );
+
+			if ( $url ) {
+				$this->register_precached_route( $url, $revision );
+				$precache_count++;
+			}
+		}
+		wp_scripts()->to_do = $original_to_do; // Restore original scripts to do.
+		return $precache_count;
+	}
+
+	/**
+	 * Register styles which are pre-cached.
+	 *
+	 * @since 0.2
+	 * @param array $handles style handles. If empty or not an array, then registered scripts with precache enabled will be used.
+	 * @return int Count of precached entries.
+	 */
+	public function register_precached_styles( $handles = array() ) {
+		if ( ! is_array( $handles ) || empty( $handles ) ) {
+			$handles = array();
+			foreach ( wp_styles()->registered as $handle => $dependency ) {
+				if ( ! empty( $dependency->extra['precache'] ) ) {
+					$handles[] = $handle;
+				}
+			}
+		}
+
+		$precache_count = 0;
+		$original_to_do = wp_styles()->to_do;
+		wp_styles()->all_deps( $handles );
+		foreach ( wp_styles()->to_do as $handle ) {
+			if ( ! isset( wp_styles()->registered[ $handle ] ) ) {
+				continue;
+			}
+			$dependency = wp_styles()->registered[ $handle ];
 
-		$routes_list = array();
-		foreach ( $routes as $route ) {
-			if ( is_string( $route ) ) {
-				$route = array( 'url' => $route );
+			// Skip bundles.
+			if ( ! $dependency->src ) {
+				continue;
 			}
-			if ( ! isset( $route['revision'] ) ) {
-				$route['revision'] = get_bloginfo( 'version' );
+
+			$url = $dependency->src;
+
+			$revision = false === $dependency->ver ? get_bloginfo( 'version' ) : $dependency->ver;
+
+			/** This filter is documented in wp-includes/class.wp-styles.php */
+			$url = apply_filters( 'style_loader_src', $url, $handle );
+
+			if ( $url ) {
+				$this->register_precached_route( $url, $revision );
+				$precache_count++;
 			}
+		}
+		wp_styles()->to_do = $original_to_do; // Restore original styles to do.
+		return $precache_count;
+	}
+
+	/**
+	 * Register runtime caching strategies for Fonts.
+	 *
+	 * @link https://developers.google.com/web/tools/workbox/guides/common-recipes#google_fonts
+	 */
+	public function register_cached_font_routes() {
+		$this->register_cached_route(
+			'^https:\/\/fonts\.(?:googleapis|gstatic)\.com\/(.*)',
+			self::STRATEGY_CACHE_FIRST,
+			array(
+				'cacheName' => 'googleapis',
+				'plugins'   => array(
+					'cacheableResponse' => array(
+						'statuses' => array( 0, 200 ),
+					),
+					'expiration'        => array(
+						'maxEntries' => 30,
+					),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Register precached routes for site icon images.
+	 *
+	 * @return int Count of pre-cached URLs.
+	 */
+	public function register_precached_site_icon() {
+		if ( ! has_site_icon() || ! get_option( 'site_icon' ) ) {
+			return 0;
+		}
 
-			$routes_list[] = $route;
+		$attachment = get_post( get_option( 'site_icon' ) );
+		if ( ! $attachment ) {
+			return 0;
 		}
-		if ( empty( $routes_list ) ) {
-			return '';
+
+		$image_urls = array_unique( array(
```

Comment:
@westonruter Why just these sizes?  Do we want to provide a filter that allows a theme developer to modify the site icon sizes, i.e. an array of sizes?  Using that strategy, we could then iterate over sizes to get the URL, register the precached route, and increment the count.

Reply:
Good question. I'll add a note explaining that the URLs here are those which are used in `wp_site_icon()`. Inside that function, there is also a `site_icon_meta_tags` filter that applies where for plugins to add new icon sizes potentially. But that will then require us to (1) do output buffering to capture, and (2) parse out the `href` attributes to obtain the URLs. There's not a clean way to just get the actual URLs. So I'll add a comment indicating that this is something we could do in the future, but for now I think this is sufficient.

Side note: it seems Chrome doesn't even respect the service worker's cache of the site icon when I go offline.

Reply:
Comment added in 4193e499b5d313cc2c944851c51867b31a642ad5